frostsnap_core/src/coordinator.rs

```diff
@@ -422,246 +423,314 @@ impl FrostCoordinator {
-                                // we use the share index as the input generator index. The input
-                                // generator at index 0 is the coordinator itself.
-                                (*share_index).into(),
-                                *response.input,
-                            )
-                            .map_err(|e| Error::coordinator_invalid_message(message_kind, e))?;
-
-                        let mut outgoing =
-                            vec![CoordinatorSend::ToUser(CoordinatorToUserMessage::KeyGen {
-                                keygen_id,
-                                inner: CoordinatorToUserKeyGenMessage::ReceivedShares { from },
-                            })];
-
-                        if input_aggregator.is_finished() {
-                            let mut agg_input = input_aggregator.clone().finish().unwrap();
-                            agg_input.grind_fingerprint::<Sha256>(Fingerprint {
-                                tag: "frostsnap-v0",
-                                bit_length: 10,
-                            });
-
-                            let contributor_keys: Vec<_> = device_to_share_index
-                                .keys()
-                                .map(|device_id| device_id.pubkey())
-                                .chain(core::iter::once(coordinator_keypair.public_key()))
-                                .collect();
-
-                            let mut certifier = certpedpop::Certifier::new(
-                                vrf_certifier.clone(),
-                                agg_input.clone(),
-                                &contributor_keys,
-                            );
-
-                            // First we calculate our (the coordinator) certificate and add our VRF outputs
-                            let sig = contributer
-                                .clone()
-                                .verify_agg_input(&vrf_certifier, &agg_input, coordinator_keypair)
-                                .unwrap();
-
-                            certifier
-                                .receive_certificate(coordinator_keypair.public_key(), sig.clone())
-                                .expect("must be able to verify our own certificate");
-
-                            outgoing.push(CoordinatorSend::ToDevice {
-                                destinations: device_to_share_index.keys().cloned().collect(),
-                                message: Keygen::CertifyPlease {
-                                    keygen_id,
-                                    agg_input: agg_input.clone(),
-                                }
-                                .into(),
-                            });
-
-                            let new_state = KeyGenState::WaitingForProofs {
-                                keygen_id,
+                use std::collections::hash_map::Entry;
```

Comment:
Cool haven't seen this pattern for avoiding lifetime issues before, very nice

Reply:
Honestly it didn't come out perfectly. Rust has IMO a missing feature here where you should be able to `.take()` from an entry and get a `Entry::Vacant` such that you can replace it later. Some discussion: https://internals.rust-lang.org/t/extend-entry-api-with-vacate-occupy/13736